### PR TITLE
OPENEUROPA-2950: Refactor theme trimming function as a Twig filter.

### PR DESCRIPTION
**Pull request by sergepavle**
### Description

[Insert description here]

### Change log

- Added:
- Changed:
- Deprecated:
- Removed:
- Fixed:
- Security:

### Commands

```sh
[Insert commands here]

```

### CODE REVIEW

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (nagyad) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: Return original text if length is not available.

File: modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php
Comment (nagyad) [inline modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php]: Please update param comment.

File: modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php
Comment (nagyad) [inline modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php]: Please update param comment.

File: modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php
Comment (nagyad) [inline modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php]: Please update parameter comment.

**Review by nagyad — CHANGES_REQUESTED** [review]

**Review by nagyad — DISMISSED** [review]

File: modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php
Comment (22Alexandra) [inline modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php]: The function needs the context for the length, not for the input data. Please set the length in the context from the data provider.

File: modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php
Comment (22Alexandra) [inline modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php]: Call the function using the $textfield from the data provider (as the function uses the parameter) instead of setting it in the $context.

**Review by 22Alexandra — CHANGES_REQUESTED** [review]

File: modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php
Comment (22Alexandra) [inline modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php]: No need to send the $context, you can send only the $length and then set it in the $context array.

**Review by 22Alexandra — CHANGES_REQUESTED** [review]

File: modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php
Comment (sergepavle) [reply modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php]: Could you explain more.

File: modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php
Comment (22Alexandra) [reply modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php]: Nevermind..better send the context already prepared with the length.

File: modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php
Comment (22Alexandra) [reply modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php]: Talked to Sergii and the result would not be the same.

**Review by 22Alexandra — DISMISSED** [review]

**Review by upchuk — DISMISSED** [review]
Not sure this is the right approach. Will check. Please don't merge.

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: Let's not use needs_context here as we don't really need it, we just need to pass the length as a parameter.

Also let's call it `smart_trim` as it's really the implementation of that behavior.

**Review by ademarco — CHANGES_REQUESTED** [review]

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: Call this 
```
   * @param int $limit
   *   Amount of text to allow.
```
As in `truncateChars()`.

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: Call the method `smartTrim()` for consistency.

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: You don't need this assignment, just return:
```
       return $truncate->truncateChars($text, $size);
```

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: This might contain in-line filters that need to be rendered before right? Like embedded media or so.

In general we should render a render array before truncating it, even if it's `#markup`, as it might have sensible `#prefix` or `#suffix` that can contain anything, including readable text.

**Review by ademarco — CHANGES_REQUESTED** [review]
I think this is not the right way of doing it, the output of that function must be a simple string, meaning that rendering should happens before trimming it.

We can also leave this to the calling template, meaning, first it must call `|render` and then `|smart_trim`, so it can be done case by case by the calling context. This would simplify this PR considerably.

Also we need to turn this into a filter, and not a function, so we can pipe it. Meaning `foo|render|smart_trim(10)`

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (sergepavle) [reply modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: done

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (sergepavle) [reply modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: done

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (sergepavle) [reply modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: fixed

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (22Alexandra) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: The $limit should be always int.

**Review by 22Alexandra — CHANGES_REQUESTED** [review]

**Review by 22Alexandra — DISMISSED** [review]

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php:528]: Let's always return a string, not at times a string and at time a markup object, I'd simply do:

```php
public function smartTrim($text, $limit = NULL) {
  // Truncate only if if length is not available or $text is string-like.
  if ($limit !== NULL && (is_string($text) || $text instanceof MarkupInterface)) {
    $truncate = new TruncateHTML();
    return $truncate->truncateChars($text, $limit);
  }

  return $text;
}
```

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: If we get a string we shouldn't turn that into a markup, as this would mean that a string with `<javascript>...</javascript>` would suddenly become executable code.

Let's simply do
```php
public function smartTrim($text, $limit) {
  $truncate = new TruncateHTML();
  return $truncate->truncateChars($text, $limit);
}
```
truncateChars handles all cases anyway, then let's have a test that ensure that passing a MarkupInterface behaves correctly.

**Review by ademarco — CHANGES_REQUESTED** [review]

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (ademarco) [reply modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: That didn't work because then the string would be stripped when rendered. We then changed approach, the only way is to respect the input.

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (brummbar) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php:526]: Can we put the type hinting on the limit?

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (brummbar) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: This comment is probably old as we don't render here anymore.

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (brummbar) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]:
```suggestion
    // If input is a Markup object, trim it and return it as such.
```

**Review by brummbar — CHANGES_REQUESTED** [review]

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php
Comment (ademarco) [reply modules/oe_theme_helper/src/TwigExtension/TwigExtension.php:526]: I'm a bit hesitant on this as it might come in a form of a template variable as a string.

Comment (brummbar) [reply modules/oe_theme_helper/src/TwigExtension/TwigExtension.php:526]: Yep indeed, let's leave it like this.

**Review by brummbar — APPROVED** [review]

**Review by ademarco — APPROVED** [review]

**Pull request merged**
